circuit/src/ecc/generic/ecdsa.rs

```diff
@@ -0,0 +1,219 @@
+		let public_key_affine = (Secp256k1::generator() * &private_key_fq).to_affine();
+		let public_key_x = Integer::from_w(public_key_affine.x.clone());
+		let public_key_y = Integer::from_w(public_key_affine.y.clone());
+		let public_key = EcPoint::new(public_key_x, public_key_y);
+		Self { private_key, public_key }
+	}
+
+	/// Generate a keypair from a given private key
+	pub fn from_private_key(private_key_fq: Fq) -> Self {
+		let private_key = Integer::from_w(private_key_fq);
+		let public_key_affine = (Secp256k1::generator() * &private_key_fq).to_affine();
+		let public_key_x = Integer::from_w(public_key_affine.x.clone());
+		let public_key_y = Integer::from_w(public_key_affine.y.clone());
+		let public_key = EcPoint::new(public_key_x, public_key_y);
+		Self { private_key, public_key }
+	}
+
+	/// Generate a signature for a given message.
+	/// Note: it does not make sense to do this in wrong field arithmetic
+	/// because the signature requires fresh randomness (k) for security reasons so it cannot be
+	/// done in a ZK circuit.
+	pub fn sign(
+		&self, msg_hash: BigUint,
+	) -> (
+		Integer<Fq, N, NUM_LIMBS, NUM_BITS, P>,
+		Integer<Fq, N, NUM_LIMBS, NUM_BITS, P>,
+	) {
+		let mut rng = thread_rng();
+		// Draw randomness
+		let k = Fq::random(rng);
+		let k_inv = k.invert().unwrap();
+
+		// Calculate `r`
+		let r_point = (Secp256k1::generator() * k).to_affine().coordinates().unwrap();
+		let x = r_point.x();
+		let r = fp_to_fq(*x);
+		let msg_hash_fq = biguint_to_fq(msg_hash);
+		let private_key_fq = biguint_to_fq(self.private_key.value());
+		// Calculate `s`
+		let s = k_inv * (msg_hash_fq + (r * private_key_fq));
+
+		(Integer::from_w(r), Integer::from_w(s))
+	}
+}
+
+/// Struct for ECDSA verification using wrong field arithmetic.
+pub struct EcdsaVerify<N: FieldExt, const NUM_LIMBS: usize, const NUM_BITS: usize, P, EC>
+where
+	P: RnsParams<<Secp256k1Affine as CurveAffine>::Base, N, NUM_LIMBS, NUM_BITS>
+		+ RnsParams<<Secp256k1Affine as CurveAffine>::ScalarExt, N, NUM_LIMBS, NUM_BITS>,
+	EC: EccParams<Secp256k1Affine>,
+{
+	phantom: std::marker::PhantomData<(N, P, EC)>,
+}
+
+impl<N: FieldExt, const NUM_LIMBS: usize, const NUM_BITS: usize, P, EC>
+	EcdsaVerify<N, NUM_LIMBS, NUM_BITS, P, EC>
+where
+	P: RnsParams<<Secp256k1Affine as CurveAffine>::Base, N, NUM_LIMBS, NUM_BITS>
+		+ RnsParams<<Secp256k1Affine as CurveAffine>::ScalarExt, N, NUM_LIMBS, NUM_BITS>,
+	EC: EccParams<Secp256k1Affine>,
+{
+	/// Verify a signature for a given message hash and public key using wrong field arithmetic
+	/// Similar to the ZK circuit setting instead of computing s_inverse we take it in as an advice value
+	pub fn verify_signature_no_pubkey_check(
+		signature: (
+			Integer<Fq, N, NUM_LIMBS, NUM_BITS, P>,
+			Integer<Fq, N, NUM_LIMBS, NUM_BITS, P>,
+		),
+		msg_hash: BigUint, public_key: EcPoint<Secp256k1Affine, N, NUM_LIMBS, NUM_BITS, P, EC>,
+		advice_s_inv: Integer<Fq, N, NUM_LIMBS, NUM_BITS, P>,
```

Comment:
Can we calculate the inverse inside the function, instead of passing it, just to have a cleaner interface?
We will do the same in circuit

Reply:
Ok, but I think this is not ideal in terms of number of circuit constraints. 

This is similar to doing division inside a circuit. When dividing `x/y` we pass in the quotient `q` as an advice and check `qy=x`.

Reply:
Done.